### PR TITLE
Fix for https://github.com/s-maheshbabu/silent-alexa/issues/22

### PR DESCRIPTION
**Pull request by s-maheshbabu**
Alexa can return her response in multiple parts. Previously, we
assumed that the parts are text and audio responses. However, as
noted in the above issue, the text itself can be broken into
multiple parts.

In the current code, since we assume that there is only one text
part, we end up showing only the first part in her multi-part
response message.

This is a preparatory commit that changes the interfaces between
ChatWindow / AVSGateway / SpeakDirectiveParser to be list of responses
instead of a single response string. There is no behavior change in
this commit besides the interface change.

### CODE REVIEW

File: src/ChatWindow.test.js [hunk 1]
```diff
@@ -293,13 +301,21 @@ const testOnUserRequestToAlexaSubmitHandling = (
...
     const finalMessages = finalState.messages;
 
-    expect(finalMessages.length).toBe(numberOfMessagesAlreadyInState + 2);
-    expect(finalMessages[finalMessages.length - 2]).toEqual(
-      expectedUserMessage
-    );
-    expect(finalMessages[finalMessages.length - 1]).toEqual(
-      expectedAlexaResponse
+    // We should have added userMessage and all of Alexa's responses.
+    const numberOfAlexaResponses = expectedAlexaResponses.size;
```
Comment (gootla-manasa) [inline src/ChatWindow.test.js]: I dont see the use of this constant. You can use expectedAlexaResponses.size wherever applicable. This is much readable than numberOfAlexaResponses which is ambiguous (can be actualAlexaResponses.size).

File: src/SpeakDirectiveParser.js [hunk 2]
```diff
@@ -1,17 +1,20 @@
...
  *
  * @param {String} alexaRawResponse The multi-part response from AVS. This
  * should not be empty or undefined.
  *
+ * @returns a List of text responses from Alexa. All the responses will be
+ * valid strings but a response can be empty if Alexa chooses to say nothing.
+ *
  * @throws IllegalArgumentError if Alexa's response couldn't be parsed out
  * of the input. The failure could be because the input is not well formatted
  * or doesn't contain the right fields to fetch Alexa's response.
```
Comment (gootla-manasa) [inline src/SpeakDirectiveParser.js]: Add a TODO here on going through the AVS documentation with scrutiny and confirm that payload empty means an error.

File: src/ChatWindow.test.js [hunk 3]
```diff
@@ -293,13 +301,21 @@ const testOnUserRequestToAlexaSubmitHandling = (
...
+    const numberOfNewMessagesInState = 1 + numberOfAlexaResponses;
+
+    expect(finalMessages.length).toBe(
+      numberOfMessagesAlreadyInState + numberOfNewMessagesInState
     );
+    expect(
```
Comment (gootla-manasa) [inline src/ChatWindow.test.js]: can you do this instead?
expect(finalMessages.get(numberOfMessagesAlreadyInState + 1)).toEqual(expectedUserMessage)
expect(finalMessages.slice(numberOfMessagesAlreadyInState + 1)).toEqual(expectedAlexaResponses)

File: src/__mocks__/AVSGateway.js [hunk 4]
```diff
@@ -1,11 +1,17 @@
-export const mockAlexaSuccessResponse = "mock alexa success response";
+const { List } = require("immutable");
```
Comment (gootla-manasa) [inline src/__mocks__/AVSGateway.js]: replace with import.

File: src/AVSGateway.js [hunk 5]
```diff
@@ -6,7 +6,7 @@ import IllegalArgumentError from "./errors/IllegalArgumentError";
 
 const uuid = require("uuid/v4");
```
[the remark this answers is not in the document]
Comment (s-maheshbabu) [reply src/AVSGateway.js]: Changed in quite a few places. Not the entire code base but related files.

File: src/ChatWindow.test.js (hunk 1 above)
Comment (s-maheshbabu) [reply src/ChatWindow.test.js]: Renamed to numberOfExpectedAlexaResponses to remove ambiguity.

File: src/ChatWindow.test.js (hunk 3 above)
Comment (s-maheshbabu) [reply src/ChatWindow.test.js]: Okay.

File: src/ChatWindow.test.js [hunk 6]
```diff
@@ -160,15 +161,20 @@ it("handles the user's form submission with request to Alexa and populates the s
...
+  let expectedAlexaResponses = List();
+  for (let alexaResponse of mockAlexaSuccessResponses) {
+    expectedAlexaResponses = expectedAlexaResponses.push(
+      new Message({
+        id: alexaId,
+        message: alexaResponse,
+        senderName: alexa.name
+      })
+    );
+  }
```
Comment (gootla-manasa) [inline src/ChatWindow.test.js]: why not use a map() function here? Something like this.
```js
const expectedAlexaResponses = mockAlexaSuccessResponses.map(  
     (alexaResponse) => new Message({
        id: alexaId,
        message: alexaResponse,
        senderName: alexa.name}));

File: src/ChatWindow.js [hunk 7]
```diff
@@ -172,11 +172,12 @@ class ChatWindow extends Component {
     // TODO: Do not make a call to avs if access_token is undefined. Redirect the user to login screen.
     avs
       .sendTextMessageEvent(userRequestToAlexa, access_token)
-      .then(response => {
-        this.pushMessage(chatterIds.ALEXA, response);
+      .then(alexaResponses => {
+        for (let alexaResponse of alexaResponses)
+          this.pushMessage(chatterIds.ALEXA, alexaResponse);
```
Comment (gootla-manasa) [inline src/ChatWindow.js]: Use map() function.
```js
alexaResponses.map((alexaResponse) => this.pushMessage(chatterIds.ALEXA, alexaResponse);

File: src/ChatWindow.test.js (hunk 6 above)
Comment (s-maheshbabu) [reply src/ChatWindow.test.js]: Done.

File: src/ChatWindow.js (hunk 7 above)
Comment (s-maheshbabu) [reply src/ChatWindow.js]: Done.

**Pull request merged**